### DISCUSSION TITLE
how to use serialport in quasar electron?

### DISCUSSION THREAD

**Discussion by s-light** [quasarframework/quasar#11433] [CLI - Electron mode]
hello community,

today i tried to revisit my [*serialport in quasar* example](https://github.com/s-light/quasar_serialport_test/)...

to just get an fresh start i tried with:
- just start with a fresh quasar project
- `yarn add electron-rebuild`
- `yarn add serialport`
- `yarn electron-rebuild -f`
- copied over my `Serial.vue` file
- adapted router and layout for new file
- `quasar dev --mode electron`
- navigat to `Serial Test`
- got a `TypeError: SerialPort.list is not a function`
-  if i just remember what i had done else to get it working back then.... :see_no_evil:
- searched a bit and [followed](https://forum.quasar-framework.org/topic/4173/can-t-use-the-serialport-module-building-for-electron) the trace to [this SO question `Quasar Framework, Electron, and Serialport: indexOf Error`](https://stackoverflow.com/questions/62360403/quasar-framework-electron-and-serialport-indexof-error) -
 i tried to replicate this - but nothing changed..
 maybe i just had no luck this day?!

if i try to import or require the serialport in the vue file i get a bunch of import related errrors during the build - 
```
App •  WAIT  • Compiling of "Renderer" in progress...
 App •  DONE  • "Renderer" compiled with errors • 415ms
 App •  ERROR  •  Renderer  in ./node_modules/@serialport/bindings/lib/darwin.js
Module not found: Can't resolve imported dependency "util"
Did you forget to install it? You can run: yarn add util
.....
```
(thsi list goes on and on and on)
seems then i try to import it in a way / context it is not meant...

- (other) things / snippets of information i found but did not manage to get combined in the correct way
    - [Electron Preload Script](https://quasar.dev/quasar-cli/developing-electron-apps/electron-preload-script)
    - [node-serialport pull request about context awareness](https://github.com/serialport/node-serialport/pull/2352)
    - [serialport install guide for electron](https://serialport.io/docs/guide-installation#electron)
    - [quasar doc: configuring-electron](https://quasar.dev/quasar-cli/developing-electron-apps/configuring-electron)

my experiments can be found in the [fresh_start branch](https://github.com/s-light/quasar_serialport_test/tree/fresh_start)

has anyone an up to date *step by step how to* for this kind of *external* module?
or any tips on how to trace this down?!
i am currently lost how to go on with this... so any tips or *broad hints* are very welcome :-)

sunny greetings
stefan

**Comment by hawkeye64** [comment]
Please provide output of `quasar info`

**Comment by s-light** [comment]
```bash
quasar_serialport_test (fresh_start)$ quasar info

Operating System - Linux(5.4.0-91-generic) - linux/x64
NodeJs - 15.14.0

Global packages
  NPM - 7.21.0
  yarn - 1.22.5
  @quasar/cli - 1.2.2
  @quasar/icongenie - 2.3.3
  cordova - Not installed

Important local packages
  quasar - 2.3.3 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app - 3.2.3 -- Quasar Framework local CLI
  @quasar/extras - 1.12.0 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.2.22 -- The progressive JavaScript framework for buiding modern web UI.
  vue-router - 4.0.12
  vuex - Not installed
  electron - 16.0.1 -- Build cross platform desktop apps with JavaScript, HTML, and CSS
  electron-packager - Not installed
  electron-builder - Not installed
  @babel/core - 7.16.0 -- Babel compiler core.
  webpack - 5.64.1 -- Packs CommonJs/AMD modules for the browser. Allows to split your codebase into multiple bundles, which can be loaded on demand. Support loaders to preprocess files, i.e. json, jsx, es7, css, less, ... and your custom stuff.
  webpack-dev-server - 4.4.0 -- Serves a webpack app. Updates the browser on changes.
  workbox-webpack-plugin - Not installed
  register-service-worker - 1.7.2 -- Script for registering service worker, with hooks
  typescript - 4.4.2 -- TypeScript is a language for application scale JavaScript development
  @capacitor/core - Not installed
  @capacitor/cli - Not installed
  @capacitor/android - Not installed
  @capacitor/ios - Not installed

Quasar App Extensions
  *None installed*

Networking
  Host - stefan-Zen
  enx00a0c6000000 - 192.168.1.4
```

**Reply by s-light** [reply]
hmpf...

i just saw the warning in the docs:
> WARNING
> Do not use uneven versions of Node i.e. 13, 15, etc. These versions are not tested with Quasar and often cause issues due to their experimental nature. We highly recommend always using the LTS version of Node.

so maybe i just run into this...

**Reply by hawkeye64** [reply]
Exactly what I was about to say...and what I was looking for in the info

**Reply by s-light** [reply]
i will try again with an updated node....
and report back here!

**Reply by s-light** [reply]
```
quasar_serialport_test (fresh_start *+)$ quasar info

Operating System - Linux(5.4.0-91-generic) - linux/x64
NodeJs - 16.13.0

Global packages
  NPM - 7.21.0
  yarn - 1.22.15
  @quasar/cli - 1.2.2
  @quasar/icongenie - 2.3.3
  cordova - Not installed

Important local packages
  quasar - 2.3.3 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app - 3.2.3 -- Quasar Framework local CLI
  @quasar/extras - 1.12.0 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.2.22 -- The progressive JavaScript framework for buiding modern web UI.
  vue-router - 4.0.12
  vuex - Not installed
  electron - 16.0.1 -- Build cross platform desktop apps with JavaScript, HTML, and CSS
  electron-packager - Not installed
  electron-builder - Not installed
  @babel/core - 7.16.0 -- Babel compiler core.
  webpack - 5.64.1 -- Packs CommonJs/AMD modules for the browser. Allows to split your codebase into multiple bundles, which can be loaded on demand. Support loaders to preprocess files, i.e. json, jsx, es7, css, less, ... and your custom stuff.
  webpack-dev-server - 4.4.0 -- Serves a webpack app. Updates the browser on changes.
  workbox-webpack-plugin - Not installed
  register-service-worker - 1.7.2 -- Script for registering service worker, with hooks
  typescript - 4.4.2 -- TypeScript is a language for application scale JavaScript development
  @capacitor/core - Not installed
  @capacitor/cli - Not installed
  @capacitor/android - Not installed
  @capacitor/ios - Not installed

Quasar App Extensions
  *None installed*

Networking
  Host - stefan-Zen
  enx00a0c6000000 - 192.168.1.4
```

- node update done.
- [migrated yarn](https://yarnpkg.com/getting-started/migration) to newest version (hopefully this was not a bad idea..)
- `yarn rebuild`
- test if serialport is working in node:
```bash
$ node
Welcome to Node.js v16.13.0.
Type ".help" for more information.
> const SerialPort = require('serialport')
undefined
> SerialPort.list().then(ports => {
...     console.log("ports:", ports.length);
...     ports.forEach(function(port) {
.....         if (port.manufacturer) {
.......             console.log(port);
.......         }
.....     });
... });
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 54,
  [Symbol(trigger_async_id_symbol)]: 43,
  [Symbol(destroyed)]: { destroyed: false }
}
> ports: 33
{
  manufacturer: 'Silicon Labs',
  serialNumber: '72ceab2d5da0eb11a3f8cbacdf749906',
  pnpId: 'usb-Silicon_Labs_CP2102N_USB_to_UART_Bridge_Controller_72ceab2d5da0eb11a3f8cbacdf749906-if00-port0',
  locationId: undefined,
  vendorId: '10c4',
  productId: 'ea60',
  path: '/dev/ttyUSB0'
}
```
- working.   
- so we should get some sort of error as this build is not compatible with electron (we have not run the electron-builder yet...)
- `quasar dev --mode electron`
  -  build all fine
  - navigate to Serial Test
  - console output:
```
listPorts()
Serial.vue?3d22:45 SerialPort ƒ SerialPort() { [native code] }
Serial.vue?3d22:46 window.serialport undefined
Serial.vue?3d22:60 SerialPort.list() failed:
 TypeError: SerialPort.list is not a function
```
  - so that is the same error i had initially.. :see_no_evil: 
- exit quasar dev
- run electron-rebuild
```bash 
$ yarn rebuild-e
⠴ Building module: bindings, Completed: 0make: Entering directory '/home/stefan/mydata/github/quasar/quasar_serialport_test/node_modules/@serialport/bindings/build'
  CXX(target) Release/obj.target/bindings/src/serialport.o
.....
make: Leaving directory '/home/stefan/mydata/github/quasar/quasar_serialport_test/node_modules/@serialport/bindings/build'
✔ Rebuild Complete
```
- `quasar dev --mode electron`
  - same as before..
- so in this case it seems  i am missing a very basic other thing...
- rechecked install instructions in [quasar](https://quasar.dev/quasar-cli/developing-electron-apps/electron-preload-script) and [serialport](https://serialport.io/docs/guide-installation#electron)
- now i get somewhere - browser console:
```
VM112 renderer_init:73 Error: Loading non-context-aware native module in renderer: '/home/stefan/mydata/github/quasar/quasar_serialport_test/node_modules/@serialport/bindings/build/Release/bindings.node'. See https://github.com/electron/electron/issues/18397.
```
- at least it tries to load :-)
- read more into this and found [this post](https://github.com/serialport/node-serialport/issues/2051#issuecomment-913727655)
> Well.. old workaround with app.allowRendererProcessReuse = false; is not working anymore with latest Electron...
> So there is no way to use serial port in electron apps right now at all ? Is there any viable workaround ?
> reconbot commented on 6 Sep
> We're actively working on a napi migration which should solve these problems. For now the latest electron may not work.

and there is a light at the end of the tunnel:
[Electron-serialport example does not work when following instructions #2166](https://github.com/serialport/node-serialport/issues/2166)

so it seems i am just a little bit to early to get this back up running ;-)

**Reply by hawkeye64** [reply]
Try with Node v14. Not sure if it'll make any difference. You can try Node v12 as well. 14 and up had significant changes. Hmm, I'm also thinking about a different version of Electron.

**Reply by s-light** [reply]
ok - maybe i was not clear enough in my findings ;-)

yeah the problem is that the currently used electron is *to new* - as it disabled the old workaround way for the `non-context-aware` modules..
so i just can wait until the good folks at the serialport package have there already existend pullrequest tested and merged...
as it is only a test setup i have time to wait :-)

thanks hawkeye64 for pointing out all the other things that can go wrong / i have to keep in mind!!

**Reply by aquiot-inc** [reply]
I struggled with this issue for days! Here is how I "solved" the problem.

I now use Quasar only for an SPA and gave up on electron integration!

For serial port communication, I have a pure Node app that wraps the serial port in an HTTP API with Web Socket for the actual receive/transmit data for the serial port. Then on the SPA client side, I call that API with Axios and client Web Socket. In `quasar dev` mode, the Axios points to my Node dev server. In `quasar build` mode, the node app also serves the html/css/js files.  I then deploy this using nexe.

While this isn't a "Quasar problem" per se (Node should have serial port, USB, BLE, etc. all built-in), the end result is that the integration of these various frameworks is very flawed!

I'm sure that there is some version of everything that worked for someone once upon a time. But serial port integration (and all other hardware communications) needs to be part of the release regression testing for the promise of Quasar to be anything more than wishful thinking. Any application framework that cannot handle simple serial port communication out-of-the-box, is a joke! (Just as much a NodeJS/Electron criticism)

I like Quasar as an SPA framework! For anything more than a web app, or something relying on Node functionality that is built-in to Node itself: forget Quasar or any other framework that tries to do the same!

There is also the fact that a simple Quasar/Electron app is  well over 100MB! But that is the consequence of the fundamental architecture that I could live with if it just worked. My Node+SPA is well under 1/3 of that size.

Anyway, I hope this will help others, I wish somebody told me this reality before I wasted so much time.

**Reply by hawkeye64** [reply]
I am not aware of any having attempted this before. While you said, it works with the backend server, but not Electron, I am curious if you tried the unbundled version vs the bundled (ASAR) version?

**Reply by aquiot-inc** [reply]
I used Quasar/Electron using `quasar build -m electron --debug`

I can make it work in debug mode using `quasar dev -m electron` after modifying `electron-main.js`
~~~ JavaScript
app.allowRendererProcessReuse = false
~~~
(per https://stackoverflow.com/questions/62360403/quasar-framework-electron-and-serialport-indexof-error)

But when building I get:
~~~ bash
Error: The module '\\?\C:\Users\dev\AppData\Local\Temp\75351054-13cb-4a28-91e5-aae18f237dc1.tmp.node' 
was compiled against a different Node.js version using 
NODE_MODULE_VERSION 83. This version of Node.js requires 
NODE_MODULE_VERSION 89. Please try re-compiling or re-installing the module 
(for instance, using `npm rebuild` or `npm install`).
~~~

And, yes: I installed `electron-rebuild` 

I tried changing versions per:
https://stackoverflow.com/questions/63794519/serialport-for-electron-node-module-version-error-and-rebuild-does-not-fix
I Tried different versions of Quasar/Node but never able to find the right combo.

I also posted this issue to https://github.com/s-light/quasar_serialport_test/issues/22  2 months ago.

and this:
https://discord.com/channels/415874313728688138/895356046393086002/895356047441657906

It seem others have struggled with this without a clear solution:
https://discord.com/channels/415874313728688138/596276102570311681/846841478143148041 

I'm not very experienced with Electron so I haven't tried unbundled version vs bundled.  With the example by @s-light broken (and remaining broken), I'm not going to be able to do any better myself without descending into a rabbit hole I really don't have time to go down.  I'm getting PTSD just thinking about it! :-)

When @s-light example (or similar) is working again (and stable), I'll consider going back to the Quasar/Electron solution. I have an API wrapper that should make it not to painful.

**Reply by hawkeye64** [reply]
Delete this file: \\?\C:\Users\dev\AppData\Local\Temp\75351054-13cb-4a28-91e5-aae18f237dc1.tmp.node
also remove node_modules and your lock file (yarn.lock or package-lock.json) and rebuild.

**Reply by aquiot-inc** [reply]
Previously tried this, and building from a clean computer. Same.

Let's see if @s-light can get it to work. Then there is hope!

**Reply by aquiot-inc** [reply]
Note that @s-light has closed the issue relating to the example, without a solution :-( 
https://github.com/s-light/quasar_serialport_test/issues/22#issuecomment-3797072242

The way I'm currently supporting serial ports is to use the [Web Serial API](https://wicg.github.io/serial/) now supported by Chrome/Edge.  However, this will only work for web apps and cannot be used for building Quasar Windows/Linux/Mac executables because the bundled browser doesn't support the Web Serial API. 

An incremental improvement would be to have the embedded browser used in Quasar support the Web Serial API (and Web USB etc.).

Ultimately, to support "Multi-platform Desktop Apps", Quasar apps need to provide a way to talk to hardware like a serial ports, USB etc. IMO.

**Reply by s-light** [reply]
yeah sorry..
i tried to cleanup my old repositories...

and in the last try i did not get anything to work..
